Add simple-nixos-mailserver to umbriel

jfly opened the pull request:
This is part 1 of https://github.com/NixOS/infra/issues/485

jfly commented on non-critical-infra/hosts/umbriel.nixos.org/mailserver/default.nix:
```diff
@@ -0,0 +1,34 @@
+    # TODO: set a `DMARC` record: <https://nixos-mailserver.readthedocs.io/en/latest/setup-guide.html#set-a-dmarc-record>.
+    #       (and then make it strict (`v=DMARC1; p=reject; adkim=s; aspf=s;`) once things are working)
+    # TODO: change to `nixos.org` when ready
+    domains = [ "playground.nixos.org" ];
```
@Mic92, I was thinking we could first set this up end to end with some other domain than `nixos.org` so we can test it out without breaking existing `nixos.org` mail. Perhaps you have a different idea for testing this out?

jfly commented on non-critical-infra/hosts/umbriel.nixos.org/mailserver/default.nix:
```diff
@@ -0,0 +1,34 @@
+  # TODO: replace with the real `nixos.org` mailing lists.
+  # TODO: port this to be a [sops-nix template](https://github.com/Mic92/sops-nix?tab=readme-ov-file#templates)
+  services.postfix.mapFiles.virtual-mailing-lists = pkgs.writeText "postfix-virtual-mailing-lists" ''
+    jfly@playground.nixos.org jfly@playground.jflei.com, jeremyfleischman+subscriber@gmail.com
```
Again, just an idea for testing this out.

Mic92 commented:
I think we only need to provide SMTP logins in special circumstances i.e. for the people that do the vote. Otherwise people can respond with their own email address for transparency reasons and use the alias in cc in case there are multiple people.

Mic92 commented:
For the sops integration, we should have an ad-hoc sops script that creates one file per recipient. Reason is that the contributor won't have access to anything but the public keys of the admin team and our server. This way they can encrypt but not decrypt.

jfly commented:
> we should have an ad-hoc sops script that creates one file per recipient

I've added a new `nix run .#encrypt-email-address` command for this. People will discover this command if they read the documentation I added in `non-critical-infra/hosts/umbriel.nixos.org/mailserver/README.md`

Mic92 commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+@click.argument("id")
+@click.argument("email")
+@click.option("--force/--no-force", "-f/ ", default=False)
+def main(id: str, email: str, force: bool):
```
```suggestion
def main(id: str, email: str, force: bool) -> None:
```

Mic92 commented on non-critical-infra/modules/mailserver/mailing-lists.nix:
```diff
@@ -0,0 +1,39 @@
+  # Encrypted email addresses go here.
+  sops.secrets.jfly-email = {
+    format = "binary";
+    sopsFile = ../../secrets/jfly-email.umbriel;
```
Maybe we want to put this in some subdirectory and import all files in it using `builtins.readDir`. Than adding new emails will be easier.

jfly replied:
I totally agree. I've reworked this. It's not quite what you suggest: instead now the list of mailing list members contains either strings or files. I then extract the files from the member lists to declare the sops secrets.

Mic92 commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+        input=email,
+    )
+
+    secret_path.write_text(cp.stdout)
```
This command would be handy:

```
  git add --intent-to-add --force -- {secret_path}
```

SuperSandro2000 commented on non-critical-infra/packages/encrypt-email-address/default.nix:
```diff
@@ -0,0 +1,24 @@
+  propagatedBuildInputs = [ python3.pkgs.click ];
+
+  installPhase = ''
+    ruff check ./encrypt-email-address.py
```
Do we really want this to potentially fail to build when ruff is being updated?

Mic92 replied:
Yes, we control the nixpkgs instance and the source here, so we should address new errors. However we already do this check in https://github.com/NixOS/infra/blob/master/formatter/flake-module.nix so this seem a bit redundant.

```suggestion
```

jfly replied:
> However we already do this check in https://github.com/NixOS/infra/blob/master/formatter/flake-module.nix so this seem a bit redundant

That's `ruff-format`, not `ruff-check`. But good point! I've added that over in https://github.com/NixOS/infra/pull/498

jfly replied:
I've removed the call to `ruff check` (in favor of #498). I'm marking this resolved.

SuperSandro2000 commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
```
```suggestion
#!/usr/bin/env python3
```

Mic92 replied:
I think this is based on suggestions of this pep: https://peps.python.org/pep-0394/

SuperSandro2000 replied:
Just to be sure no one ever tries to run the script with python2

SuperSandro2000 commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+        )
+
+    # Make sure we aren't being given a text file that happens to have a newline at the end.
+    if "\n" in email:
+        raise click.ClickException(
+            "Email address must not not contain a newline character"
+        )
```
This is kinda silly. Most editors append a final new line in linux land. Can't we rather strip whitespace?

Mic92 replied:
Isn't this email passed by cli rather than editor? But I would be also interested how we end up with a newline here.

jfly replied:
You're right @Mic92. The scenario I was vaguely worried about was somebody passing a file as a string, like this:

```
nix run .#encrypt-email-address jfly "$(< /tmp/foo)"
```

I'm also worried about the scenario where someone uses `sops edit` on these files directly, which will likely introduce a newline, and I can't think of anything clever we can do about it. Do y'all have any bright ideas?

Mic92 replied:
Mhm. Can we not just do a strip() instead to remove all whitespace around the string?

jfly replied:
Done.

SuperSandro2000 commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+            "Email address must not not contain a newline character"
+        )
+
+    if Path.cwd().name != "non-critical-infra":
+        raise click.ClickException(
+            "You must run this command inside the `non-critical-infra/` folder."
+        )
```
Just changedir to the correct directory?

Mic92 replied:
You can calculate the script directory using `Path(__name__).parent` and than compute the path you want to patch relative to that.

jfly replied:
I've implemented 1) by searching upwards for a `.git/config` directory. It was pretty simple to do: https://github.com/NixOS/infra/pull/495/commits/240fff84baebf4372d72bd4967ce30bb27ae1638. Now you can run this command from anywhere in the project =)

I'm going to mark this resolved.

Mic92 commented on non-critical-infra/modules/mailserver/mailing-lists.nix:
```diff
@@ -0,0 +1,39 @@
+  # Mailing lists go here.
+  # TODO: replace with the real `nixos.org` mailing lists.
+  lists = {
+    "jfly@playground.nixos.org" = [
```
If you haven't seen this pull request, you may not know what `playground` means. `test-mail` makes it very obvious:

```suggestion
    "jfly@test-mail.nixos.org" = [
```

jfly replied:
Updated!

Mic92 commented on non-critical-infra/modules/mailserver/default.nix:
```diff
@@ -0,0 +1,25 @@
+    # TODO: create `DKIM` TXT record: <https://nixos-mailserver.readthedocs.io/en/latest/setup-guide.html#set-dkim-signature>.
+    #       (can't do this until after SNM is deployed)
+    # TODO: set a `DMARC` record: <https://nixos-mailserver.readthedocs.io/en/latest/setup-guide.html#set-a-dmarc-record>.
+    #       (and then make it strict (`v=DMARC1; p=reject; adkim=s; aspf=s;`) once things are working)
```
```suggestion
    #       (and maybe make it strict (`v=DMARC1; p=reject; adkim=s; aspf=s;`) after checking if this is not breaking mailing lists)
```

jfly replied:
Done!

jfly reviewed (commented):
Thanks for the feedback!

jfly commented on non-critical-infra/flake-module.nix:
```diff
@@ -49,8 +50,14 @@
   perSystem =
-    { pkgs, inputs', ... }:
+    { inputs', ... }:
+    # Use the latest packages from `nixpkgs-unstable` for dev tools.
```
It's entirely non-obvious from this PR: I need sops `--filename-override` flag, which was introduced in [sops 3.9.0], which is not present in the stable nixos release we use here.

[sops 3.9.0]: https://github.com/getsops/sops/blob/main/CHANGELOG.rst#390

jfly commented on terraform/dns.tf:
```diff
@@ -330,7 +330,28 @@ locals {
+    {
+      hostname = "wiki.nixos.org"
+      type     = "TXT"
+      value    = "v=spf1 a:mail-test.nixos.org -all"
```
@Mic92, I see we're hardcoding IPs here: https://github.com/NixOS/infra/blob/fb5e9aeb6c1621c98d2c9ba029613f3be7c9e1db/terraform/dns.tf#L317, but these docs use a domain name: https://nixos-mailserver.readthedocs.io/en/latest/setup-guide.html#set-a-spf-record.

Any reason to do the IPs rather than a domain name?

Mic92 replied:
`a:` seems to be ipv4 only? We also have ipv6.

jfly replied:
Despite the name, I think it's dual ipv4/ipv6. From https://datatracker.ietf.org/doc/html/rfc7208#section-5.3:

> An address lookup is done on the <target-name> using the type of lookup (A or AAAA) appropriate for the connection type (IPv4 or IPv6)

[From experience](https://gitlab.com/simple-nixos-mailserver/nixos-mailserver/-/merge_requests/342), I know that my personal mailserver is (now) successfully sending emails to Gmail over IPv6.

jfly replied:
@mweinelt:

1. I like the thought of using `mx` rather than `a`.
2. I'm not sure about the `~all` vs `-all`. I'd be inclined to keep it as strict as possible until we have a good reason not to.

mweinelt replied:
Yeah, I prefer strictness as well.

jfly replied:
I've confirmed with my personal test mailserver that `mx` seems to work here. I've changed this SPF record to `v=spf1 mx -all`.

I'm going to mark this resolved.

jfly commented:
I'm deploying this to umbriel now

jfly commented:
@Mic92, this PR is updated with the DKIM TXT record. I've also added a README to umbriel explaining that this will need to be updated if we reprovision the machine from scratch.

jfly commented:
@Mic92, this PR is updated with the DKIM TXT record. I've also added a README to umbriel explaining that this will need to be updated if we reprovision the machine from scratch.

cole-h commented:
Just a small note: there should probably be some kind of monitoring before this sees more widespread use. Back when I ran my own instance of SNM, something (I don't remember what, it's been years at this point) went wrong and I never noticed that I stopped receiving mail (I didn't use this domain a lot at the time).

jfly commented on non-critical-infra/packages/encrypt-email-address/encrypt-email-address.py:
```diff
@@ -0,0 +1,100 @@
+@click.argument("user")
+@click.argument("email")
+@click.option("--force/--no-force", "-f/ ", default=False)
+def main(user: str, email: str, force: bool) -> None:
```
@Mic92, I don't think "user" is really the right name for this. Perhaps `address_id`?

jfly replied:
Done!

Mic92 commented:
@jfly I applied your terraform changes. Let me know when this is good to merge.

mweinelt commented:
Any reason this was just merged after #505 was opened?

jfly commented:
We wanted to get this merged because the ruff linter changes felt conflict-prone: <https://github.com/NixOS/infra/commit/8e5be8674f19dd0d38d25ef1cb64b24582e237c3>.

Mic92 commented:
@mweinelt I didn't wanted the diff get too large. And given this service is just running on mail-test.nixos.org, the issue mentioned doesn't have any real-world impact on us.
